src/cupbearer/detectors/abstraction/abstraction.py

```diff
@@ -1,198 +1,231 @@
-from __future__ import annotations  # for postponed evaluation of annotations
+from __future__ import annotations
 
-import math
-from abc import abstractmethod
-from typing import overload
+from abc import ABC, abstractmethod
+from typing import Callable
 
 import torch
+import torch.nn.functional as F
 from torch import nn
 
-from cupbearer import models
-from cupbearer.models.hooked_model import HookedModel
 
+class Abstraction(nn.Module, ABC):
+    tau_maps: _ModuleDict
 
-@overload
-def reduce_size(shape: int, size_reduction: int) -> int:
-    ...
-
-
-@overload
-def reduce_size(shape: tuple[int, ...], size_reduction: int) -> tuple[int, ...]:
-    ...
-
-
-def reduce_size(
-    shape: int | tuple[int, ...], size_reduction: int
-) -> int | tuple[int, ...]:
-    if isinstance(shape, int):
-        return math.ceil(shape / size_reduction)
-    return tuple(math.ceil(x / size_reduction) for x in shape)
+    @abstractmethod
+    def loss_fn(
+        self, name: str
+    ) -> Callable[[torch.Tensor, torch.Tensor], torch.Tensor]:
+        """Get the loss function for the given activation.
+
+        The loss function should take a prediction and a target,
+        each of shape (batch, dim), and return a tensor of shape (batch,).
+        """
+        pass
 
+    def forward(self, inputs, activations: dict[str, torch.Tensor]):
+        """Get the abstractions for the given inputs and activations.
 
-def assert_acyclic(graph: dict[str, list[str]]):
-    visited = set()
-    current_path = set()
+        Args:
+            inputs: The inputs to the model.
+            activations: A dictionary mapping activation names to the activations of the
+                full model at that name.
+        """
+        raise NotImplementedError
 
-    def visit(node):
-        if node in visited:
-            return False, None
-        visited.add(node)
-        current_path.add(node)
-        for parent in graph[node]:
-            if parent in current_path:
-                return True, current_path
-            found_cycle, cycle = visit(parent)
-            if found_cycle:
-                return True, cycle
-        current_path.remove(node)
-        return False, None
 
-    for vertex in graph:
-        found_cycle, cycle = visit(vertex)
-        if found_cycle:
-            raise ValueError(f"Cycle found in activation graph: {cycle}")
+class _ModuleDict(nn.Module):
+    """A ModuleDict that allows '.' in keys (but not '/').
 
+    Pytorch's ModuleDict doesn't allow '.' in keys, but our activation names contain
```

Comment:
The dots come from pytorch, not from us (that's how it names nested modules). I think from an end user perspective, you should be able to just pass in the names that e.g. `Module.named_modules()` gives you, rather than having to pass in names with slashes. So the main point of this dict is to hide these issues from users